R/ZIPLNfit-class.R

```diff
@@ -0,0 +1,831 @@
+          convergence = convergence)
+      )
+
+      ### TODO: Should be in post-treatment
+      if (is.null(colnames(responses))) colnames(responses) <- paste0("Y", 1:self$p)
+      colnames(private$B0) <- colnames(private$B) <- colnames(responses)
+      rownames(private$B0) <- rownames(private$B) <- colnames(covariates)
+      rownames(private$Omega)  <- colnames(private$Omega) <- colnames(private$Pi) <- colnames(responses)
+      dimnames(private$Sigma)  <- dimnames(private$Omega)
+      rownames(private$M) <- rownames(private$S) <- rownames(private$R) <- rownames(private$Pi) <- rownames(responses)
+
+    },
+
+    #' @description Result of one call to the VE step of the optimization procedure: optimal variational parameters (M, S) and corresponding log likelihood values for fixed model parameters (Sigma, B). Intended to position new data in the latent space.
+    #' @param B Optional fixed value of the regression parameters in the PLN component
+    #' @param B0 Optional fixed value of the regression parameters in the ZI component
+    #' @param Omega inverse variance-covariance matrix of the latent variables
+    #' @return A list with three components:
+    #'  * the matrix `M` of variational means,
+    #'  * the matrix `S2` of variational variances
+    #'  * the vector `log.lik` of (variational) log-likelihood of each new observation
+    optimize_vestep = function(covariates, offsets, responses, weights,
+                               B = self$model_par$B,
+                               B0 = self$model_par$B0,
+                               Omega = self$model_par$Omega,
+                               control = ZIPLN_param(backend = "nlopt")$config_optim) {
+
+      n <- nrow(responses)
+      data <- list(Y = responses, X = covariates$PLN, X0 = covariates$ZI, O = offsets)
+      parameters <-
+        list(M = matrix(0, n, self$p), S = matrix(0.1, n, self$p), R = matrix(0, n, self$p))
+
+      # Main loop
+      nb_iter <- 0
+      criterion   <- numeric(control$maxit_out)
+      convergence <- numeric(control$maxit_out)
+
+      vloglik <- -Inf; objective <- Inf
+
+      repeat {
+
+        # Check maxeval
+        if (control$maxit_out >= 0 && nb_iter >= control$maxit_out) {
+          stop_reason <- "maximum number of iterations reached"
+          criterion   <- criterion[1:nb_iter]
+          convergence <- convergence[nb_iter]
+          break
+        }
+
+        Pi <- private$optimizer$zi(
+          init_B0 = B0, X0 = data$X0, R = parameters$R, config = config_default_nlopt
+        )$Pi
+
+        # VE Step
+        new_R <- optim_zipln_R(
+          Y = data$Y, X = data$X, O = data$O, M = parameters$M, S = parameters$S, Pi = Pi
+        )
+        optim_new_M <- optim_zipln_M(
+          init_M = parameters$M,
+          Y = data$Y, X = data$X, O = data$O, R = new_R, S = parameters$S, B = B, Omega = Omega,
+          configuration = control
+        )
+        new_M <- optim_new_M$M
```

Comment:
Same as above, can we ditch `optim_new_M` and `optim_new_S` to simplify code ?

Reply:
yep.